[WebDriver] Add support for bidi session

lauromoura opened the pull request:
#### bc2351a5f32bf1139b458277a3c4cbd0ccc71322
<pre>
[WebDriver] Add support for bidi session
<a href="https://bugs.webkit.org/show_bug.cgi?id=230615">https://bugs.webkit.org/show_bug.cgi?id=230615</a>

Reviewed by Carlos Garcia Campos and BJ Burg.

Laying out the groundwork for BiDi support. This commit adds a
&quot;ENABLE_WEBDRIVER_BIDI&quot; cmake build option to use it. When enabled,
`WPEWebDriver` will spawn the libsoup3 implementation of the new
WebSocketServer in parallel to the HTTPServer of WebDriver classic.

We also added a `visibleHost` method to the HTTPServer so we can
properly build the right WebSocket URL that WPEWebDriver will forward
to the classic session. This URL is used by the webdriver client library
to connect to the right WebSocketServer address.

With this commit, we can create bidi-ready sessions using client
libraries like Selenium or WPT&apos;s wdclient, even though without support
for actual BiDi commands/events at this point.

In a future commit, we&apos;ll add actual bidi features support, like
events, starting with basic console.log messages.

* Source/WebDriver/CMakeLists.txt: Add new file.
* Source/WebDriver/CommandResult.cpp:
(WebDriver::CommandResult::httpStatusCode const): Calling new function
below.
(WebDriver::CommandResult::errorCodeToHTTPStatusCode): Added.
* Source/WebDriver/CommandResult.h:
* Source/WebDriver/HTTPServer.h:
(WebDriver::HTTPServer::visibleHost const): Added.
* Source/WebDriver/PlatformWPE.cmake: Add new file.
* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Added new constructor with reference to
the BiDi server.
(WebDriver::Session::~Session): Notify the bidi server of the classic
session closure.
* Source/WebDriver/Session.h:
(WebDriver::Session::create): Added.
(WebDriver::Session::hasBiDiEnabled const): Added.
(WebDriver::Session::setHasBiDiEnabled): Added.
* Source/WebDriver/SessionHost.cpp:
(WebDriver::browserTerminatedObservers): Added, so we can close the
WebSocket connection when the browser is terminated.
(WebDriver::SessionHost::inspectorDisconnected): Notify of browser
closure.
(WebDriver::SessionHost::addBrowserTerminatedObserver): Added.
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::WebDriverService): Instantiate BiDi server
and browser killed observer.
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run): Parse new param and instantiate
server.
(WebDriver::WebDriverService::acceptHandshake): Handle platform-agnostic
handshake algorithm steps
(WebDriver::WebDriverService::handleMessage): Placeholder, to be
implement in a future commit.
(WebDriver::WebDriverService::connectToBrowser): Add observer
(WebDriver::WebDriverService::createSession): Run extra bidi step when
creating new classic sessions
(WebDriver::WebDriverService::deleteSession): Also close BiDi session on
classic session closure.
(WebDriver::WebDriverService::clientDisconnected): Implement spec steps.
(WebDriver::WebDriverService::onBrowserTerminated): Disconnect BiDi session.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp: Added, with platform-independent
  functions.
* Source/WebDriver/WebSocketServer.h: Added.
* Source/WebDriver/glib/WebDriverServiceGLib.cpp:
(WebDriver::WebDriverService::platformSupportBidi const):
* Source/WebDriver/soup/HTTPServerSoup.cpp:
(WebDriver::HTTPServer::listen): Save visible IP for BiDi URL
* Source/WebDriver/soup/WebSocketServerSoup.cpp: Added.
* Source/cmake/WebKitFeatures.cmake: Expose ENABLE_WEBDRIVER_BIDI

Canonical link: <a href="https://commits.webkit.org/285374@main">https://commits.webkit.org/285374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68023ce618b45147a26147d88b1da0cdb9dc286e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51801 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76558 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23589 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23411 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15529 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; Passed layout tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75447 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46907 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62336 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43559 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19805 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21939 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65509 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20163 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78228 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71634 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16621 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19302 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16669 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64739 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15999 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12999 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6634 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93415 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47599 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2384 "Build is in progress. Recent messages:") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20557 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48668 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49956 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48411 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -65,13 +71,29 @@ Session::Session(std::unique_ptr<SessionHost>&& host)
     , m_pageLoadTimeout(defaultPageLoadTimeout)
     , m_implicitWaitTimeout(defaultImplicitWaitTimeout)
+#if ENABLE(WEBDRIVER_BIDI)
+    , m_bidiFlag(false)
```
Nit: use initializer in member declaration.

lauromoura replied:
Moved to use member initialization in the header.

burg commented on Source/WebDriver/Session.h:
```diff
@@ -27,12 +27,18 @@
 #include "Actions.h"
 #include "Capabilities.h"
+#include "SessionHost.h"
+#if ENABLE(WEBDRIVER_BIDI)
```
Please move the new `#if ENABLE` section to after unguarded headers.

lauromoura replied:
Moved.

burg commented on Source/WebDriver/Session.h:
```diff
@@ -236,6 +255,12 @@ class Session : public RefCounted<Session> {
     HashMap<String, InputSource> m_activeInputSources;
     HashMap<String, InputSourceState> m_inputStateTable;
+#if ENABLE(WEBDRIVER_BIDI)
+    bool m_bidiFlag;
```
What does m_bidiFlag actually signify? I think a better name here would be helpful in understanding the code. https://webkit.org/code-style-guidelines/#names-bool

Also, the capitalization is inconsistent between bidiFlag and setBiDiFlag.
https://webkit.org/code-style-guidelines/#names-basic

lauromoura replied:
Renamed to `hasBiDiEnabled`, to mean that the given `Session` has a related BiDi channel attached.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -52,12 +69,18 @@ static void printUsageStatement(const char* programName)
     String portString;
     std::optional<String> host;
+#if ENABLE(WEBDRIVER_BIDI)
+    String bidiPortString;
```
Nit: use member initializers

lauromoura replied:
Given those are local variables to the `run` method, do you mean default initializing them?

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -29,19 +29,36 @@
+#if ENABLE(WEBDRIVER_BIDI)
+#include "HTTPServer.h"
+#include "WebSocketServer.h"
+#include <algorithm>
```
I would prefer we put WTF headers above system headers.
https://webkit.org/code-style-guidelines/#include-system

lauromoura replied:
I had tried that before, but `check-webkit-style` complained about the alphabetical ordering. In fact, the following `webkitpy` test fails with the alphabetical ordering error, suggesting `<wtf..>` being treated as regular system includes:

```
        self.assert_language_rules_check('foo.h',
                                         '#include <wtf/Function.h>\n'
                                         '#include <b>\n'
                                         '#include <c>\n',
                                         '')
```

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -124,13 +155,30 @@ int WebDriverService::run(int argc, char** argv)
+#if ENABLE(WEBDRIVER_BIDI)
+    auto bidiPort = parseInteger<uint16_t>(bidiPortString);
+    if (!bidiPort) {
+        fprintf(stderr, "Invalid BiDi port %s provided. Defaulting to 4445.\n", bidiPortString.utf8().data());
```
I would call it a WebSocket port in the logging, since BiDi could mean multiple things.

lauromoura replied:
Renamed.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -124,13 +155,30 @@ int WebDriverService::run(int argc, char** argv)
+    auto bidiPort = parseInteger<uint16_t>(bidiPortString);
+    if (!bidiPort) {
+        fprintf(stderr, "Invalid BiDi port %s provided. Defaulting to 4445.\n", bidiPortString.utf8().data());
+        // TODO Make WebSocketServer listen at a random port instead of defaulting to 4445?
```
Please add a bugzilla link, or remove the comment.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,66 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    }
+
+    // 4.2 If resource name is the byte string "/session", and the implementation supports BiDi-only sessions:
+    if (*foundResource == "/session"_s) {
```
I am not sure that all these spec comments are super helpful, as spec sections will quickly diverge from this and eventually become renamed (and this code is not very complicated to begin with).

Please do keep at least the link at the top of the method, though.

lauromoura replied:
Yeah, I had added while implementing the method but now removed most of them, leaving the ones where it might add more context.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,66 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    }
+
+    // 4.4 If there is a session in the list of active sessions with session id as its session ID then let session be that session. Otherwise stop running these steps and act as if the requested service is not available.
+    // TODO Properly support multiple sessions in the future
```
Please add a bugzilla link or change the TODO to a FIXME.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,66 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    }
+
+    // 4.5 Run any other implementation-defined steps to decide if the connection should be accepted, and if it is not stop running these steps and act as if the requested service is not available.
+    // TBD
```
?

lauromoura replied:
Removed.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,66 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    // TBD
+
+    // 4.6 Otherwise append connection to session’s session WebSocket connections, and proceed with the WebSocket server-side requirements when a server chooses to accept an incoming connection.
+    // We don't have a Connection so far as we're in the middle of the handshake. Will be handled in a later step
```
Not sure this comment is helpful.

lauromoura replied:
Removed.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,66 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+void WebDriverService::handleMessage(WebSocketMessageHandler::Message&&, Function<void (WebSocketMessageHandler::Message&&)>&&)
+{
+    // https://w3c.github.io/webdriver-bidi/#handle-an-incoming-message
+    // TODO Implement message handling for actual BiDi commands
```
TODO: add bugzilla link

lauromoura replied:
Added

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -838,6 +946,10 @@ void WebDriverService::connectToBrowser(Vector<Capabilities>&& capabilitiesList,
 
     auto sessionHost = makeUnique<SessionHost>(capabilitiesList.takeLast());
+#if ENABLE(WEBDRIVER_BIDI)
+    // So we can close any living WebSocketConnection to the WebDriver client when the browser is killed.
```
Ditto about usefulness here, this is self-evident in my opinion since the variables are well named.

lauromoura replied:
Removed.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -921,6 +1036,35 @@ void WebDriverService::createSession(Vector<Capabilities>&& capabilitiesList, st
+                // 8. Set session’s BiDi flag to true.
+                m_session->setBiDiFlag(true);
+                // 9. Append "bidi" to flags.
+                // TODO
```
?

lauromoura replied:
Removed (we don't have an explicit flags member).

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -943,7 +1087,11 @@ void WebDriverService::deleteSession(RefPtr<JSON::Object>&& parameters, Function
-    session->close([session, completionHandler = WTFMove(completionHandler)](CommandResult&& result) mutable {
+    session->close([this, session, completionHandler = WTFMove(completionHandler)](CommandResult&& result) mutable {
+#if ENABLE(WEBDRIVER_BIDI)
+        // Close any bidi connections to client if they're still alive
```
Nit: unnecessary comment

lauromoura replied:
Removed.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2458,4 +2606,35 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
+void WebDriverService::onBrowserKilled(const String& sessionID)
+{
+    if (m_session && m_session->id() == sessionID) {
+        // Disconnect the WebSocket connection given session
```
Nit: end comments with '.'

lauromoura replied:
Actually removed the comment.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2458,4 +2606,35 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
+        // Disconnect the WebSocket connection given session
+        auto connection = m_bidiServer.getConnectionForSession(sessionID);
+        m_bidiServer.disconnectSession(sessionID);
+        clientDisconnected((*connection).get());
```
Is this safe to deref without a null check?

lauromoura replied:
Added check.

burg commented on Source/WebDriver/WebDriverService.h:
```diff
@@ -32,13 +32,21 @@
-class WebDriverService final : public HTTPRequestHandler {
+class WebDriverService final : public HTTPRequestHandler
+#if ENABLE(WEBDRIVER_BIDI)
+                            , public WebSocketMessageHandler
```
Please fix the spacing here per style guide:
https://webkit.org/code-style-guidelines/#punctuation-member-init

lauromoura replied:
Fixed.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+
+#include "Session.h"
+#include "WebDriverService.h"
+#include <algorithm>
```
Ditto earlier comment about system headers going last.

lauromoura replied:
Ditto about the `check-webkit-style` issue regarding `<wtf...>` and system headers.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+namespace WebDriver {
+
+WebSocketServer::WebSocketServer(WebSocketMessageHandler& messageHandler, WebDriverService& service)
+    : m_messageHandler(messageHandler), m_service(service)
```
Nit: each initializer goes on its own line.

lauromoura replied:
Fixed.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+
+bool WebSocketServer::isConnectionNotAssociatedToSession(WebSocketMessageHandler::Connection& connection)
+{
+    return std::find(m_connectionsNotAssociatedToASession.begin(), m_connectionsNotAssociatedToASession.end(), connection) != m_connectionsNotAssociatedToASession.end();
```
I think this would read better with std::count or std::count_if since we don't use the returned value.

lauromoura replied:
Replaced with `std::count`.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+{
+    m_connectionsNotAssociatedToASession.erase(std::find(m_connectionsNotAssociatedToASession.begin(), m_connectionsNotAssociatedToASession.end(), connection));
+}
+void WebSocketServer::removeConnectionAssociatedToSession(WebSocketMessageHandler::Connection& connection)
```
Nit: add newline

lauromoura replied:
Added.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+    if (errorMessage)
+        reply->setString("message"_s, *errorMessage);
+
+    // TODO Refactor httpStatusCode as a static method in CommandResult?
```
To remove?

lauromoura replied:
Removed the comment and refactored the actual code translation into a new `CommandResult::errorCodeToHTTPStatusCode`, also removing the temporary `CommandResult` used to get the code in the previous iteration.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+
+    auto serializedReply = reply->toJSONString().utf8().data();
+    auto serializedReplyLength = strlen(serializedReply);
+    return { connection, serializedReply, serializedReplyLength };
```
Nit: I think the length computation can be inlined.

lauromoura replied:
Inlined.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,234 @@
+    return command;
+}
+
+
```
Nit: extra newline, no trailing newline

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+
+#include "CommandResult.h"
+#include "HTTPServer.h"
+#include <map>
```
Ditto earlier comments about system header includes.

lauromoura replied:
Ditto `check-webkit-style` issues.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+class WebDriverService;
+
+struct Command {
+    // Defined as js-uint, but in practice opaque for the remote end
```
Nit: trailing '.'

lauromoura replied:
Added.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+// A WebSocket listener has a host, a port, a secure flag, and a list of WebSocket resources.
+struct WebSocketListener : public RefCounted<WebSocketListener> {
+    String host;
+    unsigned port;
```
Nit: use member initializers

lauromoura replied:
Added default initializers for these.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+struct WebSocketListener : public RefCounted<WebSocketListener> {
+    String host;
+    unsigned port;
+    bool secure;
```
Nit: isSecure

lauromoura replied:
Renamed.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+    }
+
+private:
+    // Constructor is private to ensure it can only be called from within the class, specifically from the create method.
```
Nit: unnecessary comment.

lauromoura replied:
Removed.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+private:
+    // Constructor is private to ensure it can only be called from within the class, specifically from the create method.
+    WebSocketListener(const String& host, unsigned port, bool secure, const std::vector<String>& resources)
+        : host(host), port(port), secure(secure), resources(resources) { }
```
Nit: initializers each on their own line.

lauromoura replied:
Fixed.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+#endif
+
+    struct Message {
+        // Optional connection, as the message might be a reply to a command not yet associated to a connection
```
This is a little bit unclear. Are you referring to commands like 'status' which are not tied to a session?

lauromoura replied:
Updated the comment for clarity. It's related to `Message` objects created while we don't have the `Connection` object available, which gets assigned in the return path of the message pipeline.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+    WebSocketMessageHandler& messageHandler() { return m_messageHandler; }
+
+    const RefPtr<WebSocketListener> listener() const { return m_listener; }
+    void addToConnectionsNotAssociatedToASession(WebSocketMessageHandler::Connection);
```
This cluster of method names is absurdly verbose and not helpful, even if it uses similar names to spec concepts. Can we just have one method for add/remove, which has an optional session argument?

lauromoura replied:
Renamed.

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+    void addToConnectionsAssociatedToASession(std::pair<WebSocketMessageHandler::Connection, String>);
+    bool isConnectionNotAssociatedToSession(WebSocketMessageHandler::Connection&);
+    void removeConnectionNotAssociatedWithSession(WebSocketMessageHandler::Connection&);
+    std::optional<RefPtr<Session>> getSessionForConnection(WebSocketMessageHandler::Connection&);
```
Naming: many of the method names are redundant with the argument names, so I think they could be shortened without losing any information.

lauromoura replied:
Renamed

burg commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,146 @@
+    WebDriverService& m_service;
+    String m_listenerURL;
+    RefPtr<WebSocketListener> m_listener;
+    // FIXME GRefPtr (our Connection type in Soup is failing to be used as a std::map key, so we use a simple pair for now)
```
Nit: add bugzilla link

lauromoura replied:
Replaced with comment about the decision to go with `std::vector` instead of `std::map` for now.

burg commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,210 @@
+        if (!response) {
+            // No error found, follow with handshake procedure
+            soup_server_unpause_message(server, message);
+        } else {
```
Nit: early return to allow dropping following block indentation.

lauromoura replied:
Replaced.

burg reviewed (changes requested):
Overall this patch is in good shape, and I have a number of smaller nits and questions. Please post a new version.

lauromoura reviewed (commented):
Thanks for the feedback, @burg! I addressed the issues (except for one rogue newline in the new file...), including renaming the "verbose" methods with more readable names and fixing the unused variable error in the windows build.

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -70,8 +76,21 @@ Session::Session(std::unique_ptr<SessionHost>&& host)
+Session::Session(std::unique_ptr<SessionHost>&& host, WeakPtr<WebSocketServer>&& bidiServer)
+    : Session(WTFMove(host))
+{
+#if ENABLE(WEBDRIVER_BIDI)
```
Uhh, we don't need two levels of #if here :)

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,53 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    // https://w3c.github.io/webdriver-bidi/#transport
+    auto resourceName = request.path;
+
+    auto resources = m_bidiServer.listener()->resources;
```
Whew, this is so much better. Thank you!

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -921,6 +1022,25 @@ void WebDriverService::createSession(Vector<Capabilities>&& capabilitiesList, st
+            // Extension steps defined by BiDi spec: https://w3c.github.io/webdriver-bidi/#establishing
+            if (!m_session->hasBiDiEnabled() && capabilities.webSocketURL && *capabilities.webSocketURL) {
+                auto listener = m_bidiServer.startListening(m_session->id());
+                // We need to update the listener host to a visible one so remote clients can connect to it
```
style nit: end comments with '.' if it's a sentence

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -943,7 +1063,11 @@ void WebDriverService::deleteSession(RefPtr<JSON::Object>&& parameters, Function
     auto session = std::exchange(m_session, nullptr);
-    session->close([session, completionHandler = WTFMove(completionHandler)](CommandResult&& result) mutable {
+    session->close([this, session, completionHandler = WTFMove(completionHandler)](CommandResult&& result) mutable {
+        UNUSED_VARIABLE(this); // Conditionally used in ENABLE(WEBDRIVER_BIDI) block.
```
Style nit: two spaces before `//` at end of the line.

burg commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+String WebSocketServer::getWebSocketURL(const RefPtr<WebSocketListener> listener, String sessionId)
+{
+    // https://w3c.github.io/webdriver-bidi/#construct-a-websocket-url
+    // FIXME: Support secure flag
```
Bug link?

burg commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+
+namespace WebDriver {
+
+static bool soupServerListen(SoupServer* server, const std::optional<String>& host, unsigned port, GError** error)
```
I'm not familiar with libsoup, so you may want another reviewer from Igalia to take a look at this file in particular for correct API usage.

carlosgcampos commented on Source/WebDriver/HTTPServer.h:
```diff
@@ -92,13 +92,16 @@ class HTTPServer
     void disconnect();
 
+    const String visibleHost() const { return m_visibleHost; }
+
```
```suggestion
    const String& visibleHost() const { return m_visibleHost; }

```

carlosgcampos commented on Source/WebDriver/Session.h:
```diff
@@ -236,6 +256,12 @@ class Session : public RefCounted<Session> {
     HashMap<String, InputSource> m_activeInputSources;
     HashMap<String, InputSourceState> m_inputStateTable;
+#if ENABLE(WEBDRIVER_BIDI)
+    bool m_hasBiDiEnabled = false;
```
```suggestion
    bool m_hasBiDiEnabled { false };
```

carlosgcampos commented on Source/WebDriver/SessionHost.cpp:
```diff
@@ -26,17 +26,33 @@
 namespace WebDriver {
 
+#if ENABLE(WEBDRIVER_BIDI)
+static WeakHashSet<SessionHost::BrowserKilledObserver>& browserKilledObservers()
```
I would better use Terminated instead of Killed

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+
+std::optional<RefPtr<Session>> WebSocketServer::getSession(WebSocketMessageHandler::Connection& connection)
+{
+    String sessionId = nullString();
```
```suggestion
    String sessionId;
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    });
+}
+
+std::optional<RefPtr<Session>> WebSocketServer::getSession(WebSocketMessageHandler::Connection& connection)
```
Why do we use optional if this returns a RefPtr? What's the difference between nullptr and nullopt in this case?

lauromoura replied:
Good catch. Dropped the `optional`

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+{
+    // https://w3c.github.io/webdriver-bidi/#start-listening-for-a-websocket-connection
+    // FIXME Add more listeners when start supporting multiple sessions
+    auto listener = m_listener;
```
Why do we need the local variable?

lauromoura replied:
Removed. Might have been a leftover from following the spec too closely.

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    return makeString("/session/"_s, sessionId);
+}
+
+String WebSocketServer::getWebSocketURL(const RefPtr<WebSocketListener> listener, String sessionId)
```
```suggestion
String WebSocketServer::getWebSocketURL(const RefPtr<WebSocketListener> listener, const String& sessionId)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    return listener;
+}
+
+String WebSocketServer::getResourceName(String sessionId)
```
```suggestion
String WebSocketServer::getResourceName(const String& sessionId)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    return sessionId;
+}
+
+RefPtr<WebSocketListener> WebSocketServer::startListening(String sessionId)
```
```suggestion
RefPtr<WebSocketListener> WebSocketServer::startListening(const String& sessionId)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    return makeString("ws://"_s, host, ":"_s, String::number(listener->port), resourceName);
+}
+
+void WebSocketServer::removeResourceForSession(const String& sessionID)
```
You are suing sessionId in the other methods, I don't mind which one, but use the same, please.

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+    String host { };
+    unsigned port { 0 };
+    bool isSecure { false };
+    std::vector<String> resources { };
```
Move these after the methods.

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+public:
+
+#if USE(SOUP)
+    typedef GRefPtr<SoupWebsocketConnection> Connection;
```
```suggestion
    using connection = GRefPtr<SoupWebsocketConnection>;
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+    struct Message {
+        // Optional connection, as the message might be generated without a connection object available (e.g. inside a method handler).
+        // In this case, it gets associated to the connection when sending the message back to the client.
+        std::optional<Connection> connection;
```
We don't need optional for this, it can just be nullptr.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2458,4 +2582,28 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
 
+#if ENABLE(WEBDRIVER_BIDI)
+
+void WebDriverService::clientDisconnected(WebSocketMessageHandler::Connection connection)
```
```suggestion
void WebDriverService::clientDisconnected(const WebSocketMessageHandler::Connection& connection)
```

carlosgcampos commented on Source/WebDriver/WebDriverService.h:
```diff
@@ -48,6 +56,8 @@ class WebDriverService final : public HTTPRequestHandler {
     static void platformInit();
     static bool platformCompareBrowserVersions(const String&, const String&);
 
+    std::optional<RefPtr<Session>> session() const { return m_session; }
```
Why optional? This never returns nullopt.

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+
+    WebSocketMessageHandler& messageHandler() { return m_messageHandler; }
+
+    const RefPtr<WebSocketListener> listener() const { return m_listener; }
```
```suggestion
    const RefPtr<WebSocketListener>& listener() const { return m_listener; }
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+{
+}
+
+void WebSocketServer::addStaticConnection(WebSocketMessageHandler::Connection connection)
```
```suggestion
void WebSocketServer::addStaticConnection(const WebSocketMessageHandler::Connection& connection)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    m_connectionToSession.push_back(connectionAndSession);
+}
+
+bool WebSocketServer::isStaticConnection(WebSocketMessageHandler::Connection& connection)
```
```suggestion
bool WebSocketServer::isStaticConnection(const WebSocketMessageHandler::Connection& connection)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    return std::count(m_staticConnections.begin(), m_staticConnections.end(), connection);
+}
+
+void WebSocketServer::removeStaticConnection(WebSocketMessageHandler::Connection& connection)
```
```suggestion
void WebSocketServer::removeStaticConnection(const WebSocketMessageHandler::Connection& connection)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    m_staticConnections.erase(std::find(m_staticConnections.begin(), m_staticConnections.end(), connection));
+}
+
+void WebSocketServer::removeConnection(WebSocketMessageHandler::Connection& connection)
```
```suggestion
void WebSocketServer::removeConnection(const WebSocketMessageHandler::Connection& connection)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+    void removeStaticConnection(WebSocketMessageHandler::Connection&);
+
+    void addConnection(std::pair<WebSocketMessageHandler::Connection, String>);
+    std::optional<RefPtr<Session>> getSession(WebSocketMessageHandler::Connection&);
```
```suggestion
    RefPtr<Session> session(const WebSocketMessageHandler::Connection&);
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+
+    void addConnection(std::pair<WebSocketMessageHandler::Connection, String>);
+    std::optional<RefPtr<Session>> getSession(WebSocketMessageHandler::Connection&);
+    std::optional<WebSocketMessageHandler::Connection> getConnection(const String& sessionId);
```
```suggestion
    WebSocketMessageHandler::Connection& connection(const String& sessionId);
```

lauromoura replied:
For this one I kept the optional, as we might not have a connection associated to the given session. Maybe return `Connection` (a `GRefPtr`) by value instead?

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,152 @@
+    String m_listenerURL;
+    RefPtr<WebSocketListener> m_listener;
+    // Soup's Connection typedef can't be used as a std::map key, so we are using a vector of pairs instead of an actual map
+    std::vector<std::pair<WebSocketMessageHandler::Connection, String>> m_connectionToSession;
```
You can use HashMap here, and Vector and any other WTF type.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+{
+    static const auto options = static_cast<SoupServerListenOptions>(0);
+
+    if (!host || host.value() == "local"_s)
```
I think host being optional makes this more difficult to read, it can be a null string if there's no host.

carlosgcampos reviewed (commented):
Added a few general comments, I'll review the soup code in a second pass.

carlosgcampos commented on Source/WebDriver/soup/HTTPServerSoup.cpp:
```diff
@@ -83,6 +83,12 @@ bool HTTPServer::listen(const std::optional<String>& host, unsigned port)
         auto* requestBody = soup_server_message_get_request_body(message);
+
+        if (httpServer.m_visibleHost.isEmpty()) {
+            auto visibleHostAndPort = soup_message_headers_get_one(soup_server_message_get_request_headers(message), "Host");
```
```suggestion
            const auto* visibleHostAndPort = soup_message_headers_get_one(soup_server_message_get_request_headers(message), "Host");
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -124,13 +156,29 @@ int WebDriverService::run(int argc, char** argv)
 
+#if ENABLE(WEBDRIVER_BIDI)
+    auto bidiURL = m_bidiServer.listen(host ? *host : nullString(), *bidiPort);
+    if (!bidiURL)
```
Why does it return a URL if we just use to check if it's nullopt or not?

lauromoura replied:
Likely a leftover from some debugging message. Removed.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +373,53 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
 
+#if ENABLE(WEBDRIVER_BIDI)
+
+void WebDriverService::handleHandshake(HTTPRequestHandler::Request&& request, Function<void (std::optional<HTTPRequestHandler::Response>&& response)>&& replyHandler)
```
Why do we need a replyHandler if this function is sync? Can't we just return Response instead?

carlosgcampos replied:
Or even return bool, since we always reply with the same 503 error, it can be handled by the caller.

lauromoura replied:
I went with returning the `Response` object, to allow other ports to reuse the same error messages/codes.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+        return;
+    }
+    GRefPtr<SoupServerMessage> protectedMessage = message;
+    soup_server_pause_message(server, message);
```
Do we really need to pause the message? This method is sync.

lauromoura replied:
Removed.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+    GRefPtr<SoupServerMessage> protectedMessage = message;
+    soup_server_pause_message(server, message);
+    auto* webSocketServer = static_cast<WebSocketServer*>(userData);
+    if (!webSocketServer) {
```
Is this really possible?

lauromoura replied:
Removed.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+        webSocketServer->addStaticConnection({ connection });
+    else {
+        auto resourceName = String::fromLatin1(path);
+        auto sessionId = webSocketServer->getSessionID(resourceName);
```
```suggestion
        auto sessionId = webSocketServer->getSessionID(String::fromLatin1(path));
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+{
+}
+
+void WebSocketServer::addStaticConnection(const WebSocketMessageHandler::Connection& connection)
```
Now trhat I see the callers this could receive a rvalue ref and move to the array.

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    m_staticConnections.push_back(connection);
+}
+
+void WebSocketServer::addConnection(const WebSocketMessageHandler::Connection& connection, const String& sessionId)
```
Same here

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+        auto resourceName = String::fromLatin1(path);
+        auto sessionId = webSocketServer->getSessionID(resourceName);
+        auto connectionForKey = WebSocketMessageHandler::Connection(connection);
+        webSocketServer->addConnection(connectionForKey, sessionId);
```
```suggestion
        webSocketServer->addConnection({ connection }, sessionId);
```

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+    g_signal_connect(connection, "message", G_CALLBACK(handleWebSocketMessage), webSocketServer);
+}
+
+
```
```suggestion
```

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+
+    gsize messageSize;
+    gconstpointer messageData = g_bytes_get_data(message, &messageSize);
+    webSocketServer->messageHandler().handleMessage({ connection, static_cast<const char*>(messageData), static_cast<size_t>(messageSize) },
```
Do you really need the size_t cast? I think g_size is a size_t

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+            return;
+        }
+    }
+}
```
Where is this used?

lauromoura replied:
This was a leftover from splitting the command dispatching parts into a follow-up patch. I've moved it there.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,208 @@
+bool WebSocketServer::isListening()
+{
+    return !!m_soupServer;
+}
```
I don't see where this is used either

lauromoura replied:
Ditto.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +372,49 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+std::optional<HTTPRequestHandler::Response> WebDriverService::handleHandshake(HTTPRequestHandler::Request&& request)
+{
+    // https://w3c.github.io/webdriver-bidi/#transport
+    auto resourceName = request.path;
```
```suggestion
    auto& resourceName = request.path;
```

carlosgcampos replied:
I guess we don't need a copy, right?

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +372,49 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+    // https://w3c.github.io/webdriver-bidi/#transport
+    auto resourceName = request.path;
+
+    auto resources = m_bidiServer.listener()->resources;
```
```suggestion
    auto& resources = m_bidiServer.listener()->resources;
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -325,6 +372,49 @@ void WebDriverService::sendResponse(Function<void (HTTPRequestHandler::Response&
+        return { { 503, "Service Unavailable", "text/plain"_s } };
+    }
+
+    return std::nullopt;
```
It's still weird that a methods called handleHandshake returns a response that is either an error or null, and when null you have to build the response yourself. If we don't want to make it bool to return the error, we can return a pair of error, message for the caller to build the response with those.

lauromoura replied:
Indeed. I might have confused with the different log messages right before creating the responses. I've renamed `handleHandshake` into `acceptHandshake` to make it clearer the purpose of the bool return. The caller (platform-specific code) now is the one creating the actual response object.

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -0,0 +1,211 @@
+    if (sessionId.isNull())
+        return { };
+
+    auto existingSession = m_service.session();
```
```suggestion
    auto& existingSession = m_service.session();
```

lauromoura replied:
Had to go with `const auto&`, as `m_service.session()` returns a rvalue and the compiler complained about assigning it to a `lvalue reference`.

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -0,0 +1,150 @@
+        return adoptRef(*new WebSocketListener(host, port, secure, resources));
+    }
+
+    String host { };
```
```suggestion
    String host;
```

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -0,0 +1,179 @@
+    gconstpointer messageData = g_bytes_get_data(message, &messageSize);
+    webSocketServer->messageHandler().handleMessage({ connection, static_cast<const char*>(messageData), messageSize },
+        [](WebSocketMessageHandler::Message&& message) {
+        if (!message.connection) {
+            WTFLogAlways("No connection found when trying to send message: %s", message.data);
+            return;
+        }
+        GRefPtr<GBytes> rawMessage = adoptGRef(g_bytes_new(message.data, message.dataLength));
+        // Using send_message to avoid dealing with null chars in the middle of the message
+        soup_websocket_connection_send_message(message.connection.get(), SOUP_WEBSOCKET_DATA_TEXT, rawMessage.get());
```
This is not correctly indented.

lauromoura replied:
According to `check-webkit-style` it's indeed 4 spaces.

For better readability, I removed the rvalue message being constructed inside the method call - that caused the like break - into a local variable, and I'm passing it with `WTFMove`. So, IIUC, the compiler can still optimize it like this initial version and we don't get the strange indent.

lauromoura committed:
[WebDriver] Add support for bidi session
https://bugs.webkit.org/show_bug.cgi?id=230615

Reviewed by Carlos Garcia Campos and BJ Burg.

Laying out the groundwork for BiDi support. This commit adds a
"ENABLE_WEBDRIVER_BIDI" cmake build option to use it. When enabled,
`WPEWebDriver` will spawn the libsoup3 implementation of the new
WebSocketServer in parallel to the HTTPServer of WebDriver classic.

We also added a `visibleHost` method to the HTTPServer so we can
properly build the right WebSocket URL that WPEWebDriver will forward
to the classic session. This URL is used by the webdriver client library
to connect to the right WebSocketServer address.

With this commit, we can create bidi-ready sessions using client
libraries like Selenium or WPT's wdclient, even though without support
for actual BiDi commands/events at this point.

In a future commit, we'll add actual bidi features support, like
events, starting with basic console.log messages.

* Source/WebDriver/CMakeLists.txt: Add new file.
* Source/WebDriver/CommandResult.cpp:
(WebDriver::CommandResult::httpStatusCode const): Calling new function
below.
(WebDriver::CommandResult::errorCodeToHTTPStatusCode): Added.
* Source/WebDriver/CommandResult.h:
* Source/WebDriver/HTTPServer.h:
(WebDriver::HTTPServer::visibleHost const): Added.
* Source/WebDriver/PlatformWPE.cmake: Add new file.
* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session): Added new constructor with reference to
the BiDi server.
(WebDriver::Session::~Session): Notify the bidi server of the classic
session closure.
* Source/WebDriver/Session.h:
(WebDriver::Session::create): Added.
(WebDriver::Session::hasBiDiEnabled const): Added.
(WebDriver::Session::setHasBiDiEnabled): Added.
* Source/WebDriver/SessionHost.cpp:
(WebDriver::browserTerminatedObservers): Added, so we can close the
WebSocket connection when the browser is terminated.
(WebDriver::SessionHost::inspectorDisconnected): Notify of browser
closure.
(WebDriver::SessionHost::addBrowserTerminatedObserver): Added.
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::WebDriverService): Instantiate BiDi server
and browser killed observer.
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run): Parse new param and instantiate
server.
(WebDriver::WebDriverService::acceptHandshake): Handle platform-agnostic
handshake algorithm steps
(WebDriver::WebDriverService::handleMessage): Placeholder, to be
implement in a future commit.
(WebDriver::WebDriverService::connectToBrowser): Add observer
(WebDriver::WebDriverService::createSession): Run extra bidi step when
creating new classic sessions
(WebDriver::WebDriverService::deleteSession): Also close BiDi session on
classic session closure.
(WebDriver::WebDriverService::clientDisconnected): Implement spec steps.
(WebDriver::WebDriverService::onBrowserTerminated): Disconnect BiDi session.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp: Added, with platform-independent
  functions.
* Source/WebDriver/WebSocketServer.h: Added.
* Source/WebDriver/glib/WebDriverServiceGLib.cpp:
(WebDriver::WebDriverService::platformSupportBidi const):
* Source/WebDriver/soup/HTTPServerSoup.cpp:
(WebDriver::HTTPServer::listen): Save visible IP for BiDi URL
* Source/WebDriver/soup/WebSocketServerSoup.cpp: Added.
* Source/cmake/WebKitFeatures.cmake: Expose ENABLE_WEBDRIVER_BIDI

Canonical link: https://commits.webkit.org/285374@main
